packages/plugin-ext/src/main/browser/webview/webview.ts

```diff
@@ -0,0 +1,212 @@
+ * Public License v. 2.0 are satisfied: GNU General Public License, version 2
+ * with the GNU Classpath Exception which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+import { BaseWidget } from '@theia/core/lib/browser/widgets/widget';
+import { IdGenerator } from '../../../common/id-generator';
+import { MiniBrowser } from '@theia/mini-browser/lib/browser/mini-browser';
+import { Disposable, DisposableCollection } from '@theia/core';
+
+export interface WebviewWidgetOptions {
+    readonly allowScripts?: boolean;
+}
+
+export interface WebviewEvents {
+    onMessage?(message: any): void;
+    onKeyboardEvent?(e: KeyboardEvent): void;
+    onLoad?(contentDocument: Document): void;
+}
+
+export class WebviewWidget extends BaseWidget {
+    private static readonly ID = new IdGenerator('webview-widget-');
+
+    protected readonly toDispose = new DisposableCollection();
+    private iframe: HTMLIFrameElement;
+    private state: string | undefined = undefined;
+    private loadTimeout: number | undefined;
+    //    private pendingMessages
+    constructor(title: string, private options: WebviewWidgetOptions, private eventDelegate: WebviewEvents) {
+        super();
+        this.id = WebviewWidget.ID.nextId();
+        this.title.closable = true;
+        // this.title.caption = this.title.label = this.props.name || 'Browser';
+        this.title.label = title;
+        // this.title.iconClass = this.props.iconClass || MiniBrowser.ICON;
+        this.addClass(MiniBrowser.Styles.MINI_BROWSER);
+
+    }
+
+    protected onFrameLoad(): void {
+        // this.hideLoadIndicator();
+        // this.focus();
+    }
+
+    protected handleMessage(message: any) {
+        switch (message.command) {
+            case 'onmessage':
+                this.eventDelegate.onMessage!(message.data);
+                break;
+            case 'do-update-state':
+                this.state = message.data;
+        }
+    }
+
+    postMessage(message: any) {
+        this.iframe.contentWindow!.postMessage(message, '*');
+    }
+
+    setHTML(html: string) {
+        html = html.replace('theia-resource:/', '/webview/');
+        html = html.replace('vscode-resource:/', '/webview/');
+        const newDocument = new DOMParser().parseFromString(html, 'text/html');
+        if (!newDocument || !newDocument.body) {
+            return;
+        }
+
+        (<any>newDocument.querySelectorAll('a')).forEach((a: any) => {
+            if (!a.title) {
+                a.title = a.href;
+            }
+        });
+
+        if (this.options.allowScripts) {
+            const defaultScript = newDocument.createElement('script');
+            defaultScript.textContent = `
+                const acquireVsCodeApi = (function() {
```

Comment:
`acquireVsCodeApi` should be provided by plugin-ext-vscode extension

Reply:
And what about `acquireTheiaApi`?

Reply:
Ok. I will add acquireTheiaApi. I am working on it

Reply:
done